File: Strivers-A2Z-DSA-Sheet/06.Bit Manipulation/3. Advanced Maths/javascript/01. Prime factors of number.js
/**
QUESTION:
Given a number N, return its unique prime factors in increasing order.

Approach:
Iterate from 2 up to sqrt(N), push every divisor that is prime, and keep dividing N by that divisor to ensure uniqueness. If the remaining N is > 1, it is also a prime factor.

Time Complexity: O(sqrt(N))
Space Complexity: O(1) aside from the output array
*/

function allPrimeFactors(n) {
    const ans = [];
    for (let i = 2; i * i <= n; i++) {
        if (n % i === 0) {
            ans.push(i);
            while (n % i === 0) {
                n = Math.floor(n / i);
            }
        }
    }
    if (n > 1) ans.push(n);
    return ans;
}

module.exports = { allPrimeFactors };

File: Strivers-A2Z-DSA-Sheet/06.Bit Manipulation/3. Advanced Maths/python/01. Prime factors of number.py
"""QUESTION:
Given an integer n, return its unique prime factors in increasing order.

Approach:
Iterate from 2 up to sqrt(n), add each divisor that divides n, and divide n repeatedly to ensure uniqueness. If the remaining n is > 1, include it.

Time Complexity: O(sqrt(n))
Space Complexity: O(1) aside from the output list
"""
from typing import List

def all_prime_factors(n: int) -> List[int]:
    ans: List[int] = []
    i = 2
    while i * i <= n:
        if n % i == 0:
            ans.append(i)
            while n % i == 0:
                n //= i
        i += 1
    if n > 1:
        ans.append(n)
    return ans
